File: backend/schemas/order_case.py
"""
OrderHub CRM — Order Case Schemas (CASE-1)

Wire shape for `/api/cases`. Two consumers: the order card (cases of one order,
with their timelines) and the dashboard block (open cases across every order the
caller may see).

The Python enums live HERE rather than on the columns. `models/order_case.py`
stores plain strings — the `Capability` precedent — so this layer is the only
thing standing between the API and a bad value, which is why every write model
below types the field as an enum rather than `str`.

NO FLOAT OR DECIMAL FIELDS ANYWHERE IN THIS MODULE, deliberately.
`tests/test_money_field_completeness.py` inspects float/Decimal response fields
and forces a revenue/cost/margin/money/neutral verdict on each. A convenience
`days_overdue: float` would drag a workflow feature that handles no money into
the money-classification surface for nothing: `due_at` is on the wire, and the
countdown is a subtraction the component can do. If a numeric ever genuinely
becomes necessary here it must be a float classified `neutral` — never a
Decimal, which serialises as a JSON string and sorts as one (`schemas/
wb_alert.py:24-31`, where `"10" < "5"` lit up a low-stock badge on a full box).
"""

import uuid
from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field

from models.order_case import OrderCaseStatus, OrderCaseType


# ─── Notes ─────────────────────────────────────────────────

class OrderCaseNoteResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    # 'comment' (a human wrote it) | 'system' (a status transition).
    kind: str
    text: str
    created_at: datetime
    # Resolved in the router for display; the column is author_id. Present on
    # system rows too — task rule 3 wants transitions attributed.
    author_id: uuid.UUID
    author_name: str | None = None


class OrderCaseNoteCreate(BaseModel):
    """Only `text`. `kind` is NOT accepted from the wire: a client must not be
    able to forge a status-transition record, which is the one row type the
    timeline is expected to trust."""

    text: str = Field(min_length=1, max_length=5000)


# ─── Cases ─────────────────────────────────────────────────

class OrderCaseResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    order_id: uuid.UUID
    case_type: str
    title: str
    status: str
    next_action: str | None = None
    due_at: datetime | None = None

    owner_id: uuid.UUID | None = None
    owner_name: str | None = None
    created_by_id: uuid.UUID
    created_by_name: str | None = None

    resolved_at: datetime | None = None
    resolution_note: str | None = None

    created_at: datetime
    updated_at: datetime

    notes: list[OrderCaseNoteResponse] = []


class OrderCaseCreate(BaseModel):
    case_type: OrderCaseType
    title: str = Field(min_length=1, max_length=500)
    next_action: str | None = Field(default=None, max_length=2000)
    due_at: datetime | None = None
    owner_id: uuid.UUID | None = None


class OrderCaseUpdate(BaseModel):
    """Every field optional — this is a PATCH.

    `status` moving to `resolved` is what stamps `resolved_at`; the service owns
    that, not the client, so there is no `resolved_at` field here. Same reason
    `notes` is absent: notes are append-only through their own endpoint.
    """

    case_type: OrderCaseType | None = None
    title: str | None = Field(default=None, min_length=1, max_length=500)
    status: OrderCaseStatus | None = None
    next_action: str | None = Field(default=None, max_length=2000)
    due_at: datetime | None = None
    owner_id: uuid.UUID | None = None
    # Only meaningful alongside `status=resolved`; ignored otherwise.
    resolution_note: str | None = Field(default=None, max_length=2000)


# ─── Dashboard ─────────────────────────────────────────────

class OpenCaseRow(BaseModel):
    """A dashboard row: the case plus just enough of its order to be clickable
    and recognisable without a second request.

    Deliberately NOT a subclass of `OrderCaseResponse`: that would carry the
    whole note timeline of every open case into the dashboard payload, and the
    block renders none of it. The row shows what a manager scans — who, what,
    what next, by when.

    No money fields — `total_price` would be `revenue`-classified and would drag
    this route onto the money surface (see the module docstring).
    """

    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    order_id: uuid.UUID
    case_type: str
    title: str
    status: str
    next_action: str | None = None
    due_at: datetime | None = None

    owner_id: uuid.UUID | None = None
    owner_name: str | None = None

    created_at: datetime

    order_number: str | None = None
    order_external_id: str | None = None
    customer_name: str | None = None
    shop_id: uuid.UUID
    shop_name: str | None = None


class OpenCasesResponse(BaseModel):
    """Two groups, matching the dashboard's two headings.

    Split server-side rather than in the component because the ordering rule
    differs per group: `in_progress` is overdue-first, `waiting` is
    due-date-first. Keeping that beside the query is what stops a second,
    drifting definition of "overdue" appearing in the frontend — the mistake
    `ParcelAlertsCard`'s docblock calls out for alerts.
    """

    # "В роботі" — overdue first, then by due date, then newest.
    in_progress: list[OpenCaseRow] = []
    # "Чекаємо" — same ordering; the ball is not in our court.
    waiting: list[OpenCaseRow] = []

File: frontend/src/types/orderCase.ts
/**
 * OrderHub CRM — Order Case types (CASE-1)
 *
 * Mirrors `backend/schemas/order_case.py`. The backend stores `case_type`,
 * `status` and `kind` as plain strings (the `Capability` precedent) and
 * validates them with Python enums at the API boundary, so the vocabulary can
 * grow without a migration. These unions are the same contract on this side —
 * and every consumer below falls back rather than crashing on a value it does
 * not recognise, because the server may grow one before the frontend ships.
 */

export const CASE_TYPES = [
  'return',
  'lost_parcel',
  'reship',
  'review',
  'address_issue',
  'claim',
  'other',
] as const

export type CaseType = (typeof CASE_TYPES)[number]

export const CASE_STATUSES = ['in_progress', 'waiting', 'resolved'] as const

export type CaseStatus = (typeof CASE_STATUSES)[number]

/** 'system' rows are written by the server on a status change. */
export type CaseNoteKind = 'comment' | 'system'

export interface OrderCaseNote {
  id: string
  kind: CaseNoteKind
  text: string
  created_at: string
  author_id: string
  author_name: string | null
}

export interface OrderCase {
  id: string
  order_id: string
  case_type: string
  title: string
  status: string
  next_action: string | null
  due_at: string | null
  owner_id: string | null
  owner_name: string | null
  created_by_id: string
  created_by_name: string | null
  resolved_at: string | null
  resolution_note: string | null
  created_at: string
  updated_at: string
  notes: OrderCaseNote[]
}

/**
 * A dashboard row. Deliberately without `notes` — the block renders none, and
 * the server does not send them.
 */
export interface OpenCaseRow {
  id: string
  order_id: string
  case_type: string
  title: string
  status: string
  next_action: string | null
  due_at: string | null
  owner_id: string | null
  owner_name: string | null
  created_at: string
  order_number: string | null
  order_external_id: string | null
  customer_name: string | null
  shop_id: string
  shop_name: string | null
}

export interface OpenCasesResponse {
  in_progress: OpenCaseRow[]
  waiting: OpenCaseRow[]
}

export interface OrderCaseCreatePayload {
  case_type: CaseType
  title: string
  next_action?: string | null
  due_at?: string | null
  owner_id?: string | null
}

export interface OrderCaseUpdatePayload {
  case_type?: CaseType
  title?: string
  status?: CaseStatus
  next_action?: string | null
  due_at?: string | null
  owner_id?: string | null
  resolution_note?: string | null
}

/** Ukrainian labels. Unknown values fall back to the raw string. */
export const CASE_TYPE_LABELS: Record<string, string> = {
  return: 'Повернення',
  lost_parcel: 'Загублена посилка',
  reship: 'Переслати',
  review: 'Відгук',
  address_issue: 'Проблема з адресою',
  claim: 'Претензія',
  other: 'Інше',
}

export const CASE_STATUS_LABELS: Record<string, string> = {
  in_progress: 'В роботі',
  waiting: 'Чекаємо',
  resolved: 'Вирішено',
}

export function caseTypeLabel(value: string): string {
  return CASE_TYPE_LABELS[value] ?? value
}

export function caseStatusLabel(value: string): string {
  return CASE_STATUS_LABELS[value] ?? value
}

/**
 * Is this case past its deadline?
 *
 * Computed here rather than sent by the server on purpose: a `days_overdue`
 * float on the wire would pull a feature that handles no money onto the
 * money-classification surface that `test_money_field_completeness.py` guards,
 * for a subtraction the browser can do. A case with no `due_at` is never
 * overdue — an undated case is not late.
 */
export function isOverdue(dueAt: string | null, now: Date = new Date()): boolean {
  if (!dueAt) return false
  return new Date(dueAt).getTime() < now.getTime()
}
